From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Fri, 17 Feb 2023 16:31:50 -0800
Subject: [PATCH] Optimize map id lookups

Micro optimisation to avoid a double contains lookup on the CompoundTag

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 29192b350cb8638acf2c35480eb4ce603523f31e..1d3d712b4b83f0867bc5b7eb3ccde2a388717dae 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -109,6 +109,11 @@ public class PlutoConfig {
         return (List<T>) config.getList(path, config.getList(path));
     }
 
+    public static boolean useMapIdLookupOptimisation = true;
+    private static void shouldUseMapIdLookupOptimisation() {
+        useMapIdLookupOptimisation = getBoolean("items.maps.use-id-lookup-optimisation", useMapIdLookupOptimisation);
+    }
+
     public static final class WorldConfig {
 
         public final String worldName;
diff --git a/src/main/java/net/minecraft/world/item/MapItem.java b/src/main/java/net/minecraft/world/item/MapItem.java
index d3c29e6bf8b3c2dd628809177dac50220a7de415..edf3d82216579262fd2625bab92cdfc25cd6471b 100644
--- a/src/main/java/net/minecraft/world/item/MapItem.java
+++ b/src/main/java/net/minecraft/world/item/MapItem.java
@@ -73,6 +73,16 @@ public class MapItem extends ComplexItem {
     public static Integer getMapId(ItemStack stack) {
         CompoundTag nbttagcompound = stack.getTag();
 
+        // Pluto start - Optimize map id lookups
+        if (dev.yive.pluto.PlutoConfig.useMapIdLookupOptimisation) {
+            if (nbttagcompound == null) return null;
+
+            net.minecraft.nbt.Tag tag = nbttagcompound.get("map");
+            if (tag == null) return null;
+
+            return tag instanceof net.minecraft.nbt.NumericTag map ? map.getAsInt() : 0;
+        }
+        // Pluto end - Optimize map id lookups
         return nbttagcompound != null && nbttagcompound.contains("map", 99) ? nbttagcompound.getInt("map") : null; // CraftBukkit - make new maps for no tag // Paper - don't return invalid ID
     }
 
